### PR TITLE
Creates a BYO Sink Example Guide

### PR DESCRIPTION
**Pull request by JeffNeff**
Fixes issue https://github.com/knative/docs/issues/4203

## Proposed Changes 

- Creates a guide to building basic event sinks.

### CODE REVIEW

**Review by cab105 — COMMENTED** [review]
Overall, I like the guide you provided as it does clarify the other half of the Knative pipe. I left some feedback for your review.

File: docs/developer/eventing/sinks/creating-event-sinks/writing-event-sinks-with-golang/README.md
Comment (JeffNeff) [reply docs/developer/eventing/sinks/creating-event-sinks/writing-event-sinks-with-golang/README.md:185]: no. it will build as is.

File: docs/developer/eventing/sinks/creating-event-sinks/writing-event-sinks-with-golang/README.md
Comment (JeffNeff) [reply docs/developer/eventing/sinks/creating-event-sinks/writing-event-sinks-with-golang/README.md:137]: as an example. should we not?

Comment (cab105) [reply docs/developer/eventing/sinks/creating-event-sinks/writing-event-sinks-with-golang/README.md:137]: KISS unless there's something special you want to highlight, and that should be documented.

File: docs/developer/eventing/sinks/creating-event-sinks/writing-event-sinks-with-golang/README.md
Comment (JeffNeff) [reply docs/developer/eventing/sinks/creating-event-sinks/writing-event-sinks-with-golang/README.md:17]: I personally like all the code inline. but I am okay with it living next to this document if that is what the consensus is.

File: docs/developer/eventing/sinks/creating-event-sinks/writing-event-sinks-with-golang/README.md
Comment (JeffNeff) [reply docs/developer/eventing/sinks/creating-event-sinks/writing-event-sinks-with-golang/README.md]: I updated it but I like the demonstration of testing locally being there as well. what do you think of this? 
https://github.com/knative/docs/pull/4208/commits/04c094601726b4d7bd31b25ff0bdcfda23b7f722

Comment (cab105) [reply docs/developer/eventing/sinks/creating-event-sinks/writing-event-sinks-with-golang/README.md]: I'm with you on that. Although for consistency, it should be up around L180 as well as L298.

Comment (JeffNeff) [reply docs/developer/eventing/sinks/creating-event-sinks/writing-event-sinks-with-golang/README.md]: im with you on that... how about it all goes in one place? https://github.com/knative/docs/pull/4208/commits/63fbc06568d2be1b934a2f7bfde3f636fb33a62f

File: config/nav.yml
Comment (abrennan89) [inline config/nav.yml:109]:
```suggestion
            - Writing an event sink using Golang: developer/eventing/sinks/creating-event-sinks/writing-event-sinks-with-golang/README.md
```

Comment (abrennan89) [reply config/nav.yml:109]: I think this should also go below Kafka - "About sinks" should stay the first topic in the list, please

File: docs/developer/eventing/sinks/creating-event-sinks/writing-event-sinks-with-golang/README.md
Comment (abrennan89) [inline docs/developer/eventing/sinks/creating-event-sinks/writing-event-sinks-with-golang/README.md:7]: this frontmatter isn't required anymore since me moved from hugo to mkdocs

File: docs/developer/eventing/sinks/creating-event-sinks/writing-event-sinks-with-golang/README.md
Comment (abrennan89) [inline docs/developer/eventing/sinks/creating-event-sinks/writing-event-sinks-with-golang/README.md:10]: please write "Knative Service" rather than "ksvc" to avoid confusion

File: docs/developer/eventing/sinks/creating-event-sinks/writing-event-sinks-with-golang/README.md
Comment (abrennan89) [inline docs/developer/eventing/sinks/creating-event-sinks/writing-event-sinks-with-golang/README.md:12]: Use sentence case for all headings:
"A quick overview of building and deploying a simple event sink with Golang"

File: docs/developer/eventing/sinks/creating-event-sinks/writing-event-sinks-with-golang/README.md
Comment (abrennan89) [inline docs/developer/eventing/sinks/creating-event-sinks/writing-event-sinks-with-golang/README.md:14]:
```suggestion
In the following procedure, you can create a simple sink that receives CloudEvents and performs an example operation on them.
```

File: docs/developer/eventing/sinks/creating-event-sinks/writing-event-sinks-with-golang/README.md
Comment (abrennan89) [inline docs/developer/eventing/sinks/creating-event-sinks/writing-event-sinks-with-golang/README.md:16]: Make these numbered steps please so that it's easier for folks to follow

Comment (abrennan89) [inline docs/developer/eventing/sinks/creating-event-sinks/writing-event-sinks-with-golang/README.md:16]:
```suggestion
1. From within the `$GOPATH`, create a folder named `demosink`, that contains a new file named `main.go`:
```

File: docs/developer/eventing/sinks/creating-event-sinks/writing-event-sinks-with-golang/README.md
Comment (abrennan89) [inline docs/developer/eventing/sinks/creating-event-sinks/writing-event-sinks-with-golang/README.md:160]:
```suggestion
1. Create the `go.mod` file, then run the following commands:
```
If you do add numbered steps to the doc, please indent the code samples so that the numbering renders properly also.

File: docs/developer/eventing/sinks/creating-event-sinks/writing-event-sinks-with-golang/README.md
Comment (abrennan89) [inline docs/developer/eventing/sinks/creating-event-sinks/writing-event-sinks-with-golang/README.md:176]:
```suggestion
1. Create a new file `Dockerfile` in the same folder:
    
    ```dockerfile
    FROM golang:1.15-buster AS builder
    ENV CGO_ENABLED 0
    WORKDIR /project
    COPY . ./
    RUN go build
    FROM gcr.io/distroless/static:nonroot
    COPY --from=builder /project/demosink /
    ENTRYPOINT ["/demosink"]
    ```
```

suggestion of what I meant wrt. numbering and indentation 🙂

Comment (abrennan89) [reply docs/developer/eventing/sinks/creating-event-sinks/writing-event-sinks-with-golang/README.md:176]: repeat for all below steps...let me know if you have any trouble; I can just add a clean up PR after this merges if you do.

File: docs/developer/eventing/sinks/creating-event-sinks/writing-event-sinks-with-golang/README.md
Comment (abrennan89) [inline docs/developer/eventing/sinks/creating-event-sinks/writing-event-sinks-with-golang/README.md:217]: All of our documentation should be vendor neutral, so if this sample requires Google platform specific tools etc please modify it so that there is a generic version instead, e.g. using `kubectl` commands.

**Review by abrennan89 — CHANGES_REQUESTED** [review]
Added some suggestions and comments. From a docs perspective there would be quite a bit of cleanup to do here; if you have any issues with that let me know and I can submit a PR to clean up after you have merged this.

I think the only blocking thing I see is the use of the `gcloud` CLI - our docs should all be vendor neutral, so any samples etc included must not require these types of platform specific tools. For instance, we would also not allow use of OpenShift's `oc` CLI; `kubectl` should be used where possible.

File: config/nav.yml
Comment (JeffNeff) [reply config/nav.yml]: fixed : https://github.com/knative/docs/pull/4208/commits/972319f71e0fdb34a8f18f1ab41ce013e72d4c53

Thank you!

File: docs/developer/eventing/sinks/creating-event-sinks/writing-event-sinks-with-golang/README.md
Comment (JeffNeff) [reply docs/developer/eventing/sinks/creating-event-sinks/writing-event-sinks-with-golang/README.md]: fixed: https://github.com/knative/docs/pull/4208/commits/b6cadbe876d44a25e58478fb8732f39eee14f712
Thank you!

File: docs/developer/eventing/sinks/creating-event-sinks/writing-event-sinks-with-golang/README.md
Comment (JeffNeff) [reply docs/developer/eventing/sinks/creating-event-sinks/writing-event-sinks-with-golang/README.md:153]: Okay! Thank you!

File: docs/developer/eventing/sinks/creating-event-sinks/writing-event-sinks-with-golang/README.md
Comment (JeffNeff) [reply docs/developer/eventing/sinks/creating-event-sinks/writing-event-sinks-with-golang/README.md]: Fixed: https://github.com/knative/docs/pull/4208/commits/b8225598113892d3f55f56b0d5d006485761e575 

Thank you!

File: docs/developer/eventing/sinks/creating-event-sinks/writing-event-sinks-with-golang/README.md
Comment (JeffNeff) [reply docs/developer/eventing/sinks/creating-event-sinks/writing-event-sinks-with-golang/README.md]: Fixed: https://github.com/knative/docs/pull/4208/commits/b8225598113892d3f55f56b0d5d006485761e575 
Thank you!

File: docs/developer/eventing/sinks/creating-event-sinks/writing-event-sinks-with-golang/README.md
Comment (JeffNeff) [reply docs/developer/eventing/sinks/creating-event-sinks/writing-event-sinks-with-golang/README.md]: Fixed: https://github.com/knative/docs/pull/4208/commits/6f43db24c1ef811a54d9de91acf0bcc1f2c9aa68 
Thank you!

File: docs/developer/eventing/sinks/creating-event-sinks/writing-event-sinks-with-golang/README.md
Comment (JeffNeff) [reply docs/developer/eventing/sinks/creating-event-sinks/writing-event-sinks-with-golang/README.md]: Fixed by removing the `gcloud` reference: https://github.com/knative/docs/pull/4208/commits/891e4c9ba7c5f72755deebdae93b81a2c7866c73 
Thank you!

**Comment by JeffNeff** [comment]
> Added some suggestions and comments. From a docs perspective there would be quite a bit of cleanup to do here; if you have any issues with that let me know and I can submit a PR to clean up after you have merged this.
> 
> I think the only blocking thing I see is the use of the `gcloud` CLI - our docs should all be vendor neutral, so any samples etc included must not require these types of platform specific tools. For instance, we would also not allow use of OpenShift's `oc` CLI; `kubectl` should be used where possible.

Okay! Thank you for your time in reviewing this! I have no issues with anything moving forward :)

**Comment by mattray** [comment]
I ran through the steps and was able to recreate everything.

**Comment by abrennan89** [comment]
@JeffNeff fantastic, thank you SO much 🙂 
/lgtm
/approve
/hold

I'll leave a hold on this so you can remove it to merge when you're ready, since I see it's still marked as `Draft` state.

**Comment by abrennan89** [comment]
Actually might want to get an additional technical review from someone in @knative/eventing-wg-leads or the #eventing Slack channel just to be sure 🙂

**Comment by csantanapr** [comment]
Is this a duplicate sample as the one we have here? https://github.com/knative/docs/blob/main/docs/serving/samples/cloudevents/cloudevents-go/cloudevents.go

**Pull request closed without merging**
